src/anemoi/graphs/generate/icon_mesh.py

```diff
@@ -0,0 +1,395 @@
+# (C) Copyright 2024 Anemoi contributors.
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+import itertools
+import logging
+import uuid
+from dataclasses import dataclass
+from functools import cached_property
+from typing import Optional
+
+import netCDF4
+import numpy as np
+import scipy
+from typeguard import typechecked
+from typing_extensions import Self
+
+from anemoi.graphs.generate.utils import convert_adjacency_matrix_to_list
+from anemoi.graphs.generate.utils import convert_list_to_adjacency_matrix
+from anemoi.graphs.generate.utils import selection_matrix
+
+LOGGER = logging.getLogger(__name__)
+
+
+@typechecked
+class NodeSet:
+    """Stores nodes on the unit sphere."""
+
+    id_iter: int = itertools.count()  # unique ID for each object
+    gc_vertices: np.ndarray  # geographical (lat/lon) coordinates [rad], shape [:,2]
+
+    def __init__(self, lon: np.ndarray, lat: np.ndarray):
+        self.gc_vertices = np.column_stack((lon, lat))
+        self.id = uuid.uuid4()
+
+    @property
+    def num_vertices(self) -> int:
+        return self.gc_vertices.shape[0]
+
+    @cached_property
+    def cc_vertices(self):
+        """Cartesian coordinates [rad], shape [:,3]."""
+        return self._gc_to_cartesian()
+
+    def __add__(self, other: Self) -> Self:
+        """concatenates two node sets."""
+        gc_vertices = np.concatenate((self.gc_vertices, other.gc_vertices))
+        return NodeSet(gc_vertices[:, 0], gc_vertices[:, 1])
+
+    def __eq__(self, other: Self) -> bool:
+        """Compares two node sets."""
+        return self.id == other.id
+
+    def _gc_to_cartesian(self, radius: float = 1.0) -> np.ndarray:
+        """Returns Cartesian coordinates of the node set, shape [:,3]."""
+        xyz = (
+            radius * np.cos(lat_rad := self.gc_vertices[:, 1]) * np.cos(lon_rad := self.gc_vertices[:, 0]),
+            radius * np.cos(lat_rad) * np.sin(lon_rad),
+            radius * np.sin(lat_rad),
+        )
+        return np.stack(xyz, axis=-1)
+
+
+@typechecked
+@dataclass
+class EdgeID:
+    """Stores additional categorical data for each edge (IDs for heterogeneous input)."""
+
+    edge_id: np.ndarray
+    num_classes: int
+
+    def __add__(self, other: Self):
+        """Concatenates two edge ID datasets."""
+        assert self.num_classes == other.num_classes
+        return EdgeID(
+            edge_id=np.concatenate((self.edge_id, other.edge_id)),
+            num_classes=self.num_classes,
+        )
+
+
+@typechecked
+class GeneralGraph:
+    """Stores edges for a given node set."""
+
+    nodeset: NodeSet  # graph nodes
+    edge_vertices: np.ndarray  # vertex indices for each edge, shape [:,2]
```

Comment:
It might be more consistent to call them "edge_indices" as this is the terminology used in the other edge building components.

Reply:
Hi Helen, thanks for your review. Would you agree to keep this refactoring for a follow-up PR? I quickly grepped through `anemoi-*` and could not find anything for `edge_indices`. 
Just in [PyG](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.EdgeIndex.html#torch_geometric.EdgeIndex) I found a definition of `EdgeIndex`. However they  transposed the index order. I think that aligning the index order would definitely beyond the scope of this PR.

However, if you prefer using `edge_indices`, I'd be more than happy to grant you permission to make the necessary changes to this branch. Please do feel free and refactor as needed to get this PR over the line.